python/spacetime/managers/version_manager.py

```diff
@@ -404,36 +409,50 @@ def receive_data(self, appname, versions, package, from_external=True):
         if start_v == end_v:
             # The versions are the same, lets ignore.
             return True
-        if start_v != self.version_graph.head.current:
+        if start_v!= self.version_graph_head:
+        #if start_v != self.version_graph.head.current:
             self.resolve_conflict(start_v, end_v, package, from_external)
         else:
             self.version_graph.continue_chain(start_v, end_v, package)
+            self.version_graph_head=end_v
         self.maintain(appname, end_v)
         return True
 
     def retrieve_data(self, appname, version):
+
         data, version_change = self.retrieve_data_nomaintain(version)
         self.set_app_marker(appname, version_change[1])
         return data, version_change
 
+    #def retrieve_data_nomaintain(self, version):
+        #merged = dict()
+        #if version not in self.version_graph.nodes:
+            #return merged, [version, version]
+        #for delta in self.version_graph[version:]:
+            #merged = utils.merge_state_delta(merged, delta)
+        #return merged, [version, self.version_graph.head.current]
+
     def retrieve_data_nomaintain(self, version):
         merged = dict()
-        if version not in self.version_graph.nodes:
-            return merged, [version, version]
+        
         for delta in self.version_graph[version:]:
             merged = utils.merge_state_delta(merged, delta)
-        return merged, [version, self.version_graph.head.current]
+        if not merged:
```

Comment:
You also need to make the equivalent check in Graph class in __getitem__ to return an empty list if the start version does not exist.